File: matrix_java/src/M.java
public class M {
    int rows; // 矩阵行数
    int cols; // 矩阵列数
    F[][] data; // 存储矩阵数据的二维数组

    public M(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.data = new F[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                set(i, j, 0);
            }
        }
    }

    public M(long[][] data) {
        this.rows = data.length;
        this.cols = data[0].length;
        this.data = new F[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                set(i, j, data[i][j]);
            }
        }
    }

    public M(F[][] data) {
        this.rows = data.length;
        this.cols = data[0].length;
        this.data = data;
    }

    public F[][] getData() {
        return data;
    }

    // 获取矩阵中某个位置的值
    public F get(int i, int j) {
        return data[i][j];
    }


    // 设置矩阵中某个位置的值
    public void set(int i, int j, F val) {
        data[i][j] = val;
    }

    // 设置矩阵中某个位置的值
    public void set(int i, int j, long val) {
        set(i, j, new F(val));
    }

    public M add(M B) {
        M A = this;
        if (A.rows != B.rows || A.cols != B.cols) {
            throw new IllegalArgumentException("The matrixs can not add");
        }
        int n = A.rows;
        int m = A.cols;
        M C = new M(n, m);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                C.set(i, j, A.get(i, j).add(B.get(i, j)));
            }
        }
        return C;
    }

    public M sub(M B) {
        M A = this;
        if (A.rows != B.rows || A.cols != B.cols) {
            throw new IllegalArgumentException("The matrixs can not sub");
        }
        int n = A.rows;
        int m = A.cols;
        M C = new M(n, m);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                C.set(i, j, A.get(i, j).sub(B.get(i, j)));
            }
        }
        return C;
    }

    public M multiply(M B) {
        M A = this;
        if (A.cols != B.rows) {
            throw new IllegalArgumentException("The matrixs can not multiply");
        }
        int n = A.rows;
        int m = B.cols;
        int p = A.cols;
        M C = new M(n, m);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                F sum = new F(0);
                for (int k = 0; k < p; k++) {
                    sum = sum.add(A.get(i, k).multiply(B.get(k, j)));
                }
                C.set(i, j, sum);
            }
        }
        return C;
    }

    public M pow(long n) {
        M res = new M(rows, cols);
        res.eye();
        M it = this;
        if (n < 0) {
            it = it.inverse();
            n = n * -1;
        }
        while (n != 0) {
            if (n % 2 == 1) {
                res = res.multiply(it);
            }
            it = it.multiply(it);
            n /= 2;
        }
        return res;
    }

    public M divie(M B) {
        M A = this;
        M invB = B.inverse();
        if (A.cols != invB.rows) {
            throw new IllegalArgumentException("The matrixs can not divide");
        }
        return this.multiply(invB);
    }

    public M inverse() {
        M A = this;
        // 判断矩阵是否为方阵
        if (A.rows != A.cols) {
            throw new IllegalArgumentException("The matrix is not square");
        }
        int n = A.rows;
        M E = new M(n, n);
        // 初始化为单位矩阵
        E.eye();
        // 初始化为A的一个拷贝
        M U = this.copy();
        // 高斯-约旦消元
        for (int k = 0; k < n; k++) {
            F max = U.abs(k, k);
            int pivot = k;
            //找绝对值最大的行
            for (int i = k + 1; i < n; i++) {
                if (U.abs(i, k).compare(max) > 0) {
                    max = U.abs(i, k);
                    pivot = i;
                }
            }
            //如果绝对值最大为0，则无解
            if (max.len() == 0) {
                System.out.println("原矩阵无法下一步操作");
                System.out.println(U);
                throw new RuntimeException("Matrix is not invertible");
            }
            // 交换行
            if (k != pivot) {
                U.swapRow(pivot, k);
                E.swapRow(pivot, k);
            }
            // 将主元归一
            F divisor = U.get(k, k);
            for (int j = 0; j < n; j++) {
                U.set(k, j, U.get(k, j).divide(divisor));
                E.set(k, j, E.get(k, j).divide(divisor));
            }
            // 消元
            for (int i = 0; i < n; i++) {
                if (i != k) {
                    F factor = U.get(i, k);
                    for (int j = 0; j < n; j++) {
                        U.set(i, j, U.get(i, j).sub(factor.multiply(U.get(k, j))));
                        E.set(i, j, E.get(i, j).sub(factor.multiply(E.get(k, j))));
                    }
                }
            }
        }
        return E;
    }


    public F determinant() {
        M A = this;
        // 判断矩阵是否为方阵
        if (A.rows != A.cols) {
            throw new IllegalArgumentException("The matrix is not square");
        }
        int n = A.rows;
        // 初始化为A的一个拷贝
        M U = this.copy();
        // 高斯-约旦消元
        F det = new F(1);
        for (int k = 0; k < n; k++) {
            F max = U.abs(k, k);
            int pivot = k;
            //找绝对值最大的行
            for (int i = k + 1; i < n; i++) {
                if (U.abs(i, k).compare(max) > 0) {
                    max = U.abs(i, k);
                    pivot = i;
                }
            }
            //如果绝对值最大为0，则行列式为0
            if (max.len() == 0) {
                return max;
            }
            // 交换行
            if (k != pivot) {
                U.swapRow(pivot, k);
                det = det.multiply(new F(-1));
            }
            F divisor = U.get(k, k);
            det = det.multiply(divisor);
            // 将主元归一
            for (int j = 0; j < n; j++) {
                U.set(k, j, U.get(k, j).divide(divisor));
            }
            // 消元
            for (int i = k + 1; i < n; i++) {
                F factor = U.get(i, k);
                for (int j = 0; j < n; j++) {
                    U.set(i, j, U.get(i, j).sub(factor.multiply(U.get(k, j))));
                }
            }
        }
        return det;
    }

    public M solve() {
        M A = this;
        int n = A.rows;
        int m = A.cols;
        // 拷贝一份。
        M U = A.copy();
        // 高斯-约旦消元
        for (int k = 0, T = 0; k < n && T < m; k++, T++) {
            F max = U.abs(k, T);
            int pivot = k;
            // 找绝对值最大的行
            for (int i = k + 1; i < n; i++) {
                if (U.abs(i, T).compare(max) > 0) {
                    max = U.abs(i, T);
                    pivot = i;
                }
            }
            // 如果绝对值最大为0，则无解
            if (max.len() == 0) {
                k--;
                continue;
            }
            // 交换行
            if (k != pivot) {
                U.swapRow(pivot, k);
            }
            // 将主元归一
            F divisor = U.get(k, T);
            for (int j = 0; j < m; j++) {
                U.set(k, j, U.get(k, j).divide(divisor));
            }
            // 消元
            for (int i = 0; i < n; i++) {
                if (i != k) {
                    F factor = U.get(i, T);
                    for (int j = 0; j < m; j++) {
                        U.set(i, j, U.get(i, j).sub(factor.multiply(U.get(k, j))));
                    }
                }
            }
        }
        return U;
    }

    public M inversepro() {
        M A = this;
        int n = A.rows;
        int m = A.cols;
        // 初始化为A的一个拷贝
        M U = this.copy();
        //特殊单位矩阵
        M E = new M(n, m);
        E.eye();
        // 高斯-约旦消元
        for (int k = 0, T = 0; k < n && T < m; k++, T++) {
            F max = U.abs(k, T);
            int pivot = k;
            //找绝对值最大的行
            for (int i = k + 1; i < n; i++) {
                if (U.abs(i, T).compare(max) > 0) {
                    max = U.abs(i, T);
                    pivot = i;
                }
            }
            //如果绝对值最大为0，则无解
            if (max.len() == 0) {
                k--;
                continue;
            }
            // 交换行
            if (k != pivot) {
                U.swapRow(pivot, k);
                E.swapRow(pivot, k);
            }
            // 将主元归一
            F divisor = U.get(k, T);
            for (int j = 0; j < m; j++) {
                U.set(k, j, U.get(k, j).divide(divisor));
                E.set(k, j, E.get(k, j).divide(divisor));
            }
            // 消元
            for (int i = 0; i < n; i++) {
                if (i != k) {
                    F factor = U.get(i, T);
                    for (int j = 0; j < m; j++) {
                        U.set(i, j, U.get(i, j).sub(factor.multiply(U.get(k, j))));
                        E.set(i, j, E.get(i, j).sub(factor.multiply(E.get(k, j))));
                    }
                }
            }
        }
        return E;
    }

    public M transpose() {
        M A = this;
        int n = A.rows;
        int m = A.cols;
        M B = new M(m, n);
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                B.set(i, j, A.get(j, i));
            }
        }
        return B;
    }

    public void swapRow(int a, int b) {
        int n = cols;
        for (int j = 0; j < n; j++) {
            F temp = this.get(a, j);
            this.set(a, j, this.get(b, j));
            this.set(b, j, temp);
        }
    }

    public void swapCol(int a, int b) {
        int n = rows;
        for (int j = 0; j < n; j++) {
            F temp = this.get(j, a);
            this.set(j, a, this.get(j, b));
            this.set(j, b, temp);
        }
    }

    public void eye() {
        M e = this;
        int n = e.rows;
        int m = e.cols;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                e.set(i, j, 0);
            }
            e.set(i, i, 1);
        }
    }

    public M copy() {
        M e = new M(rows, cols);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                e.set(i, j, this.get(i, j));
            }
        }
        return e;
    }

    public F abs(int i, int j) {
        F f = this.get(i, j);
        if (f.compare(new F()) < 0) {
            f = f.multiply(new F(-1));
        }
        return f;
    }


    public String toString() {
        return print();
    }

    // 打印矩阵
    public String print() {
        int maxLen = 0;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols - 1; j++) {
                maxLen = Math.max(maxLen, String.valueOf(data[i][j]).length());
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < rows; i++) {
            StringBuilder line = new StringBuilder();
            for (int j = 0; j < cols; j++) {
                int key = maxLen + 1;
                String str = String.format("%-" + key + "s", data[i][j]);
                line.append(str).append(" ");
            }
            line.append("\n");
            sb.append(line);
        }
        return sb.toString();
    }
}

File: matrix_python/M.py
from copy import copy

from C import Complex


class M:
    def __init__(self, *args):
        if len(args) == 1 and isinstance(args[0], list):
            # 假设第一个参数是二维列表
            data = args[0]
            self.row = len(data)
            self.col = len(data[0])
            if self.row > 0 and self.col > 0 and not isinstance(data[0][0], Complex):
                temp = [[Complex() for _ in range(self.col)] for _ in range(self.row)]
                for i in range(self.row):
                    for j in range(self.col):
                        temp[i][j] = Complex(data[i][j])
                data = temp
            self.data = data
        elif len(args) == 2 and all(isinstance(arg, int) for arg in args):
            # 假设有两个参数，且都是整数
            row, col = args
            self.row = row
            self.col = col
            self.data = [[Complex(0) for _ in range(col)] for _ in range(row)]
        else:
            raise ValueError("Invalid arguments to M constructor")

    def __getitem__(self, key):
        if isinstance(key, tuple):
            # 如果key是元组，则直接返回对应的元素
            return self.data[key[0]][key[1]]
        else:
            # 否则返回一个可以再次被索引的行对象
            return self.data[key]

    def __setitem__(self, key, value):
        if isinstance(key, tuple):
            # 如果key是元组，则设置对应元素的值
            self.data[key[0]][key[1]] = value
        else:
            # 如果不是元组，那么就是设置整行的值
            self.data[key] = value

    def __add__(self, B):
        if not isinstance(B, M):
            B = M(B)
        A = M(self.data)
        if A.row != B.row or A.col != B.col:
            raise ValueError("row or col not unequal")
        n = A.row
        m = A.col
        C = M(n, m)
        for i in range(n):
            for j in range(m):
                C[i][j] = A[i][j] + B[i][j]
        return C

    def __sub__(self, B):
        if not isinstance(B, M):
            B = M(B)
        A = M(self.data)
        if A.row != B.row or A.col != B.col:
            raise ValueError("row or col not unequal")
        n = A.row
        m = A.col
        C = M(n, m)
        for i in range(n):
            for j in range(m):
                C[i][j] = A[i][j] - B[i][j]
        return C

    def __mul__(self, B):
        if not isinstance(B, M):
            B = M(B)
        A = M(self.data)
        if A.col != B.row:
            raise ValueError("can not multiply")
        n = A.row
        m = B.col
        l = A.col
        C = M(n, m)
        for i in range(n):
            for j in range(m):
                sum = Complex(0)
                for k in range(l):
                    sum += A[i][k] * B[k][j]
                C[i][j] = sum
        return C

    def __iadd__(self, other):
        if not isinstance(other, M):
            B = M(other)
        return self + other

    def __isub__(self, other):
        if not isinstance(other, M):
            B = M(other)
        return self - other

    def __imul__(self, other):
        if not isinstance(other, M):
            B = M(other)
        return self * other

    def __pow__(self, power, modulo=None):
        n = power
        bas = M(self.data)
        res = self.eye(self.row)
        while n > 0:
            if n % 2 == 1:
                res *= bas
            bas *= bas
            n //= 2
        return res

    def __copy__(self):
        n = self.row
        m = self.col
        it = [[self.data[0][0] for _ in range(m)] for _ in range(n)]
        for i in range(n):
            for j in range(m):
                it[i][j] = self.data[i][j]
        return M(it)

    def eye(self, n: int):
        E = M(n, n)
        for i in range(n):
            E[i][i] = Complex(1)
        return E

    def inverse(self):
        A = M(self.data)
        if A.row != A.col:
            raise ValueError("can not inverse")
        n = A.row
        U = copy(A)
        E = self.eye(n)

        for r in range(n):
            mx = U[r][r]
            pivot = r
            for i in range(r + 1, n):
                if abs(U[i][r]) > abs(mx):
                    mx = U[i][r]
                    pivot = i
            if mx == Complex(0):
                raise ValueError("can not inverse")
            if r != pivot:
                U[pivot], U[r] = U[r], U[pivot]
                E[pivot], E[r] = E[r], E[pivot]
            divisor = U[r][r]
            for c in range(n):
                U[r][c] //= divisor
                E[r][c] //= divisor
            for i in range(n):
                if i == r:
                    continue
                factor = U[i][r]
                for j in range(n):
                    U[i][j] -= factor * U[r][j]
                    E[i][j] -= factor * E[r][j]
        return E

    def solve(self):
        A = M(self.data)
        n = A.row
        m = A.col

        U = copy(A)
        r = 0
        c = 0
        while r < n and c < m:
            mx = U[r][c]
            pivot = r
            for i in range(r + 1, n):
                if abs(U[i][c]) > abs(mx):
                    mx = U[i][c]
                    pivot = i
            if mx == Complex(0):
                c += 1
                continue
            if r != pivot:
                U[pivot], U[r] = U[r], U[pivot]
            divisor = U[r][c]
            for i in range(m):
                U[r][i] //= divisor
            for i in range(n):
                if i == r:
                    continue
                factor = U[i][c]
                for j in range(m):
                    U[i][j] -= factor * U[r][j]
            r += 1
            c += 1
        return U

    def __str__(self):
        maxLen = 0
        for row in self.data:
            for col in row:
                maxLen = max(len(str(col)), maxLen)
        res = ""
        maxLen += 1
        for row in self.data:
            for col in row:
                res += "{:<{width}} ".format(str(col), width=maxLen)
            res += '\n'
        return res

    def __neg__(self):
        T = copy(self)
        for i in range(T.row):
            for j in range(T.col):
                T.data[i][j] = -T.data[i][j]
        return T


if __name__ == "__main__":
    A = M([[0, 1], [2, 3]])
    B = M([[Complex(1, 9), Complex(5, 2)], [Complex(7, 3), Complex(4, 9)]])
    C = A.inverse()
    print(A + B)
    print(A - B)
    print(A * B)
    print(B.inverse())
    print(A * B.inverse())
    D = M([
        [Complex(1), Complex(2), Complex(0), Complex(-2), Complex(-4)],
        [Complex(3, 9), Complex(5), Complex(1), Complex(-5), Complex(-11)],
        [Complex(3), Complex(-2, 2), Complex(8), Complex(3), Complex(-0)],
        [Complex(2), Complex(-3), Complex(7), Complex(4, 8), Complex(3)],
    ])
    print(D.solve())
